feat(system-server): add GET /system/connected

fsinapi opened the pull request:
# Overview

This PR adds in a route `GET /system/connected` that will return information about all of the currently connected clients. In order to implement this, an `AuthorizationTracker` class keeps track of the clients that obtain a token through the `POST /system/authorize` route.

Closes RCORE-520 and RCORE-515

Requires #12198, will change the base branch to `edge` once that is merged

# Test Plan

Tested on a Dev server & through integration testing:

- Register a client, then ping /system/connected and ensure _nothing_ pops up (registration does not count as connecting)
- Authorize that client, ping /system/connected, ensure that the correct Registrant info shows up
- Authorize the client again, ping /system/connected, ensure the same info comes back (no duplication)
- Register & authorize a second client, ping /system/connected, ensure _both_ sets of Registration info come back correctly

# Changelog

- Added `AuthorizationTracker` class that maintains a list of authorizations & their expiration datetimes
- Added `expiration_from_jwt` function to extract the expiration time from a JWT
- Added functionality in `POST /system/authorize` to add authorizations to a singleton tracker
- Added `GET /system/connected` route that extracts authorizations from the singleton tracker
- Modified the dev server fixture to run a new instance for each test - this increases test time a bit, but establishes a clean server state after each test.

# Review requests


# Risk assessment

Low

fsinapi committed:
Add code to generate a JWT

fsinapi committed:
Added functionality to verify a JWT is valid

fsinapi committed:
Added HTTP endpoint to register a new JWT
* Added database integration for saving/loading/deleting a JWT
* Added tests for database integration
* Added placeholder implementation of the HTTP endpoint POST /system/register

fsinapi committed:
Fixed an issue with sql syntax and added a more thorough test for database integration

fsinapi committed:
Removed fake testing endpoint
* Added a proper model for the response to POST /system/register

fsinapi committed:
added Tavern to system server dev packages

fsinapi committed:
Added a Tavern integration test for /system/register

fsinapi committed:
Take out an unnecessary line in dev_server.py

fsinapi committed:
Some cleanup

fsinapi committed:
Moved a constant to constant.py

fsinapi committed:
Removed unnecessary async defs

fsinapi committed:
Moved some core logic from the /system/register router to make it actually testable, and added a corresponding test.

fsinapi committed:
Addressed review comments:
* Moved JWT algorithm into a constant
* Evaluate current time when function is called, not when Python module is loaded
* Removed sleep call during system server wakeup

fsinapi committed:
Added optional ID field to jwt creator

fsinapi committed:
Added an HTTP header to check for a Registration Token

fsinapi committed:
Added POST /system/authorize endpoint

fsinapi committed:
Updated /system/register to return code 201 if the JWT is new

fsinapi committed:
Merge branch 'RCORE-509_system_register_route' into RCORE-510_post_system_authorize

fsinapi committed:
Fixed a test to use the correct HTTP response code

fsinapi committed:
Fixed an inaccurate comment

fsinapi committed:
Merge branch 'RCORE-509_system_register_route' into RCORE-510_post_system_authorize

fsinapi committed:
Linting changes

fsinapi committed:
Update system-server/system_server/system/register/storage.py

Co-authored-by: Max Marrone <max@opentrons.com>

fsinapi committed:
Simplify registration storage tests

fsinapi committed:
Issue all registration database commands in a single transaction

fsinapi committed:
Fixed a misleading function comment

fsinapi committed:
Renamed `get_uuid` to be more clear that this is a persistent value

fsinapi committed:
Removed unnecessary try block

fsinapi committed:
Added documentation for both 200 and 201 response codes

fsinapi committed:
Moved Registration parameters into a dependency with documentation

fsinapi committed:
Added more detail to system_jwt.py public functions

fsinapi committed:
Merge branch 'edge' into RCORE-509_system_register_route

fsinapi committed:
Merge branch 'RCORE-509_system_register_route' into RCORE-510_post_system_authorize

# Conflicts:
#	system-server/system_server/jwt/system_jwt.py

fsinapi committed:
Updates to work with RCORE-509

fsinapi committed:
Merge branch 'edge' into RCORE-510_post_system_authorize

# Conflicts:
#	system-server/system_server/constants.py
#	system-server/system_server/jwt/__init__.py
#	system-server/system_server/jwt/system_jwt.py
#	system-server/system_server/system/router.py
#	system-server/tests/jwt/test_jwt.py

fsinapi committed:
Added class for tracking active connections

fsinapi committed:
Expanded integration test to check that subsequent /system/authorize calls don't return the same token

fsinapi committed:
Fixed typo in a comment

fsinapi committed:
changed authentication_bearer parameter to be camelCase

fsinapi committed:
Added another test for active tracking

fsinapi committed:
Moved active tracker to its own folder; added a function to get expiration from a jwt

fsinapi committed:
Added authorization tracking to POST /system/authorize

fsinapi committed:
Added GET /system/connected and corresponding integration test

fsinapi committed:
There's no reason for the authorization tracker to have to be async - there weren't any yields!

fsinapi committed:
Renamed the active tracker file to authorization_tracker

fsinapi committed:
Merge upstream changes

sfoster1 commented on system-server/system_server/connection/authorization_tracker.py:
```diff
@@ -0,0 +1,58 @@
+        # If this registrant already exists, remove it
+        for c in self._connections:
+            if c.registrant == registrant:
+                self._connections.remove(c)
```
this alters `self._connections` while iterating through it which is not a good idea. Maybe put the found value in your pocket and remove it after the end of the iteration, or do a filtered list copy

sfoster1 commented on system-server/system_server/connection/authorization_tracker.py:
```diff
@@ -0,0 +1,58 @@
+        now = datetime.now()
+        for c in self._connections:
+            if c.expiration < now:
+                self._connections.remove(c)
```
same deal - this alters `_connections` while iterating through it, may be better as `self._connections = [c for c in self._connections if c.expiration > now]`

sfoster1 commented on system-server/system_server/jwt/system_jwt.py:
```diff
@@ -38,6 +38,19 @@ def registrant_from_jwt(token: str, signing_key: str) -> Registrant:
+        options={"verify_aud": False},
+    )
+    exp = decoded["exp"]
+    print(f"Timestamp: {exp}")
```
should be a `log` or removed

sfoster1 commented on system-server/system_server/jwt/system_jwt.py:
```diff
@@ -38,6 +38,19 @@ def registrant_from_jwt(token: str, signing_key: str) -> Registrant:
+    )
+    exp = decoded["exp"]
+    print(f"Timestamp: {exp}")
+    return datetime.fromtimestamp(float(exp))
```
`float` doesn't seem right here, shouldn't this be an int?

sfoster1 commented on system-server/system_server/persistence/__init__.py:
```diff
@@ -29,6 +33,7 @@
 _persistence_dir_lock = Lock()
 _sql_lock = Lock()
 _uuid_lock = Lock()
+_authorization_tracker_lock = Lock()
```
at some point we should refactor so these locks are held in application state

SyntaxColoring replied:
Yep, this has been bothering me in `robot-server` too. It's a good candidate for some kind of reusable util.

sfoster1 reviewed (changes requested):
Structure looks great, couple nits around altering collections while iterating through them

fsinapi committed:
Don't alter a list while iterating through it

fsinapi committed:
Addressed review comments
- Removed a debugging print statement
- Convert datetime from int, not float
- Left a TODO about some refactor work that's a little too involved to toss into this PR

sfoster1 reviewed (approved):
Looks good to me!

fsinapi committed:
Merge branch 'edge' into RCORE-520_active_connection_tracking

# Conflicts:
#	system-server/system_server/jwt/__init__.py
#	system-server/system_server/jwt/system_jwt.py
#	system-server/system_server/service/check_jwt_headers.py
#	system-server/system_server/system/authorize/router.py
#	system-server/system_server/system/router.py
#	system-server/tests/integration/test_system_authorize.tavern.yaml
#	system-server/tests/jwt/test_jwt.py
